### PR TITLE
Some command line argument parsing with clap

### LINKED ISSUE #3: Add support for command line options (clap)
Add support for command line options using the [clap](https://github.com/clap-rs/clap) library. 

- [ ] add `clap` as a dependency
- [ ] port tockloader's command line arguments to clap
- [ ] when running, display the list of supplied arguments

### PR DESCRIPTION
**Pull request by eva-cosma**
In this Pull Request, I've implemented some basic command line argument interpretation using [clap](https://github.com/clap-rs/clap). This is in no way a full translation from the python version, just some basic building blocks. 

The python version uses two base categories of "arguments":
- app arguments, which are shared by most subcommands which interact with the apps
- channel arguments, which configure the computer-board communication

Some subcommands do have additional arguments.

I've implemented the two categories (with the exception of the deprecated "jtag" arguments), and I've made a simple proof of concept, to test that clap is indeed parsing correctly. 

Some screenshots:
image
image
image

P.S. I attempted to implement the command line arguments with "derive", but I did not manage to find a way for subcommands to inherit arguments. This would've meant explicitly typing out all of the arguments for all of the subcommands.

### CODE REVIEW

File: rust/Cargo.lock [hunk 1]
```diff
@@ -2,6 +2,237 @@
 # It is not intended for manual editing.
 version = 3
 
+[[package]]
```
Comment (alexandruradovici) [inline rust/Cargo.lock]: Please exclude this file from the repository.

File: rust/src/cli.rs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+/// Generate all of the [arguments](clap::Arg) that are required by subcommands which work
+/// with channels and computer-board communication.
+fn get_channel_args() -> Vec<clap::Arg> {
+    vec![
+        arg!(-p --port "The serial port or device name to use"),
+        arg!(--serial "Use the serial bootloader to flash")
+            .action(clap::ArgAction::SetTrue),
+        // Should we include depracated functionality?
```
Comment (alexandruradovici) [inline rust/src/cli.rs]: I don't think so.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruradovici** [comment]
> P.S. I attempted to implement the command line arguments with "derive", but I did not manage to find a way for subcommands to inherit arguments. This would've meant explicitly typing out all of the arguments for all of the subcommands.

I don't understand this, can you show us an example?

**Comment by eva-cosma** [comment]
> > P.S. I attempted to implement the command line arguments with "derive", but I did not manage to find a way for subcommands to inherit arguments. This would've meant explicitly typing out all of the arguments for all of the subcommands.
> 
> I don't understand this, can you show us an example?

Sure! 
To create a clap Command, we have the option to either declare the command, with all of its components (subcommands, arguments, etc. ), using a builder or using structs and enums.

The difference between the two can be seen in clap's own examples:
- [git.rs](https://github.com/clap-rs/clap/blob/master/examples/git.rs) for the first method (code result [here](https://github.com/clap-rs/clap/blob/master/examples/git-derive.md))
- [git_derive.rs](https://github.com/clap-rs/clap/blob/master/examples/git-derive.rs) for the second method (code result [here](https://github.com/clap-rs/clap/blob/master/examples/git.md))

In my own opinion, the derive method is way more straight-forward and cleaner. However, when attempting to learn this method I had come upon an issue when trying to port from the python version.

As I explained before, a lot of subcommands share a lot of arguments. To solve this, the original authors grouped the common arguments into a single "parser" object (one parser object for the apps, and one for the "channel"/board-computer communication). From there, individual subcommands are given 0,1, or both parsers as "parents", inheriting their arguments. From there, the subcommands can then define their own specific arguments.

Here is a code snippet from main.py where the "app" parser is defined: [click me](https://github.com/UPB-CS-OpenSourceUpstream/tockloader/blob/74d37580a4d8f19d2e16a508f1a4bd80023551aa/python/tockloader/main.py#L553)
And here is where the the parser is then used, in the "install" command: [click me](https://github.com/UPB-CS-OpenSourceUpstream/tockloader/blob/74d37580a4d8f19d2e16a508f1a4bd80023551aa/python/tockloader/main.py#L761)

Through my experimentation, I have not managed to find a way to replicate this behavior using the "derive" method.

**Comment by alexandruradovici** [comment]
Thanks for the explanation, what I do not clearly understand is how you use command arguments *inheritance* right now?

**Comment by eva-cosma** [comment]
The way I did that was to set up two different functions that will build a vector of arguments, which can then be added to each individual command.

Right now, the "listen" subcommand is set up to use both the "app" and "channel" arguments.

```rs
    vec![Command::new("listen")
        .about("Open a terminal to receive UART data")
        .args(get_app_args()) // Use the APP arguments
        .args(get_channel_args()) // Use the CHANNEL arguments
        .arg_required_else_help(true)]
```

The two functions are then defined as such:
```rs
fn get_app_args() -> Vec<clap::Arg> {
    vec![<REMOVED FOR BERVITY>]
}

fn get_channel_args() -> Vec<clap::Arg> {
    vec![<REMOVED FOR BREVITY>]
}
```

If in the future if we want to implement a subcommand which uses only the app arguments, and some subcommand-specific arguments, then we can create the new subcommand like this:

```rs
Command::new("my_command")
        .about("My awesome subcommand")
        .args(get_app_args()) // Use the APP arguments
        .args(get_my_command_specific_args())  // Arguments that are only used by this command. (can be defined inline, without helper function)
]
```

This isn't exactly inheritance, but it does allow re-using the same arguments for multiple subcommands.

File: rust/src/cli.rs (hunk 2 above)
Comment (eva-cosma) [reply rust/src/cli.rs]: I've removed the comment and the deprecated argument

**Comment by alexandruradovici** [comment]
Ok, got it.

**Comment by alexandruradovici** [comment]
Closes #3

**Comment by alexandruradovici** [comment]
@george-cosma Please select a timeslot for the interview https://doodle.com/meeting/participate/id/eZ8gOOEb.

**Pull request merged**